File: shared/src/main/java/ru/privatenull/pnauth/config/ProxySettings.java
package ru.privatenull.pnauth.config;

import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public record ProxySettings(
        boolean requireServerAuth,
        String authServer,
        List<String> authServers,
        String backendServer,
        List<String> backendServers,
        Map<String, String> forcedHosts
) {
    public ProxySettings(boolean requireServerAuth, String authServer) {
        this(requireServerAuth, authServer, List.of(), "hub", List.of(), Map.of());
    }

    public ProxySettings(boolean requireServerAuth, String authServer, String backendServer, Map<String, String> forcedHosts) {
        this(requireServerAuth, authServer, List.of(), backendServer, List.of(), forcedHosts);
    }

    public ProxySettings {
        authServer = normalize(authServer);
        backendServer = normalize(backendServer);
        authServers = normalizeList(authServers);
        backendServers = normalizeList(backendServers);
        forcedHosts = forcedHosts == null ? Map.of() : Map.copyOf(forcedHosts);

        if (requireServerAuth && allAuthServers(authServer, authServers).isEmpty()) {
            throw new IllegalArgumentException(
                    "Authentication routing is enabled, but no authentication server is configured. "
                            + "Set servers.auth-server or servers.auth-servers, or disable servers.require-auth-before-server.");
        }

        Set<String> auth = lowerCaseSet(allAuthServers(authServer, authServers));
        for (String backend : allBackendServers(backendServer, backendServers)) {
            if (auth.contains(backend.toLowerCase(java.util.Locale.ROOT))) {
                throw new IllegalArgumentException(
                        "Authentication and backend routes must use different server names. Conflicting server: '"
                                + backend + "'. Check servers.auth-server(s) and servers.backend-server(s).");
            }
        }

        if (forcedHosts.entrySet().stream().anyMatch(entry -> entry.getKey() == null || entry.getKey().isBlank()
                || entry.getValue() == null || entry.getValue().isBlank())) {
            throw new IllegalArgumentException("servers.forced-hosts must not contain blank hostnames or server names");
        }
    }

    public boolean hasBackendServer() {
        return !backendServerCandidates().isEmpty();
    }

    public List<String> authServerCandidates() {
        return allAuthServers(authServer, authServers);
    }

    public List<String> backendServerCandidates() {
        return allBackendServers(backendServer, backendServers);
    }

    public boolean isAuthServer(String serverName) {
        if (serverName == null || serverName.isBlank()) return false;
        return authServerCandidates().stream().anyMatch(serverName::equalsIgnoreCase);
    }

    public boolean isBackendServer(String serverName) {
        if (serverName == null || serverName.isBlank()) return false;
        return backendServerCandidates().stream().anyMatch(serverName::equalsIgnoreCase);
    }

    public static ProxySettings defaults() {
        return new ProxySettings(true, "auth", List.of(), "hub", List.of(), Map.of());
    }

    public ProxySettings requiringServerAuth() {
        return requireServerAuth ? this
                : new ProxySettings(true, authServer, authServers, backendServer, backendServers, forcedHosts);
    }

    private static String normalize(String value) {
        return value == null ? "" : value.trim();
    }

    private static List<String> normalizeList(List<String> values) {
        if (values == null || values.isEmpty()) return List.of();
        LinkedHashSet<String> result = new LinkedHashSet<>();
        for (String value : values) {
            String normalized = normalize(value);
            if (!normalized.isEmpty()) result.add(normalized);
        }
        return List.copyOf(result);
    }

    private static List<String> allAuthServers(String primary, List<String> additional) {
        return merge(primary, additional);
    }

    private static List<String> allBackendServers(String primary, List<String> additional) {
        return merge(primary, additional);
    }

    private static List<String> merge(String primary, List<String> additional) {
        LinkedHashSet<String> result = new LinkedHashSet<>();
        if (primary != null && !primary.isBlank()) result.add(primary.trim());
        if (additional != null) {
            for (String value : additional) {
                if (value != null && !value.isBlank()) result.add(value.trim());
            }
        }
        return List.copyOf(new ArrayList<>(result));
    }

    private static Set<String> lowerCaseSet(List<String> values) {
        Set<String> result = new LinkedHashSet<>();
        for (String value : values) result.add(value.toLowerCase(java.util.Locale.ROOT));
        return result;
    }
}

File: shared/src/main/kotlin/ru/privatenull/pnauth/config/ProxySettings.kt
package ru.privatenull.pnauth.config

import ru.privatenull.pnauth.routing.ServerBalancerMode

/** One routable proxy server with its configured online capacity. */
class ServerTarget @JvmOverloads constructor(
    @JvmField var server: String = "",
    @JvmField var maxOnline: Int = 100
)

data class ProxySettings(
    val requireServerAuth: Boolean,
    val authServers: List<ServerTarget>,
    val backendServers: List<ServerTarget>,
    val forcedHosts: Map<String, String> = emptyMap(),
    val balancerMode: ServerBalancerMode = ServerBalancerMode.LEAST_PLAYERS
) {
    /** Compatibility constructor for extensions compiled against the previous routing model. */
    @JvmOverloads
    constructor(
        requireServerAuth: Boolean,
        authServer: String,
        backendServer: String = "hub",
        forcedHosts: Map<String, String> = emptyMap(),
        backendServers: List<String> = if (backendServer.isNotBlank()) listOf(backendServer) else emptyList(),
        authServers: List<String> = if (authServer.isNotBlank()) listOf(authServer) else emptyList(),
        balancerMode: ServerBalancerMode = ServerBalancerMode.LEAST_PLAYERS,
        maxPlayersPerServer: Int = 100,
        serverLimits: Map<String, Int> = emptyMap()
    ) : this(
        requireServerAuth = requireServerAuth,
        authServers = effectiveNames(authServer, authServers).map {
            ServerTarget(it, serverLimits[it] ?: maxPlayersPerServer)
        },
        backendServers = effectiveNames(backendServer, backendServers).map {
            ServerTarget(it, serverLimits[it] ?: maxPlayersPerServer)
        },
        forcedHosts = forcedHosts,
        balancerMode = balancerMode
    )

    init {
        if (requireServerAuth && authServers.isEmpty()) {
            throw IllegalArgumentException(
                "Authentication routing is enabled, but servers.auth-servers is empty. " +
                        "Add at least one auth server or disable servers.require-auth-before-server."
            )
        }
        if (authServers.any { it.server.isBlank() } || backendServers.any { it.server.isBlank() }) {
            throw IllegalArgumentException("servers.auth-servers/backend-servers must not contain blank server names")
        }
        if (authServers.any { it.maxOnline <= 0 } || backendServers.any { it.maxOnline <= 0 }) {
            throw IllegalArgumentException("Every server max-online value must be greater than 0")
        }
        val authNames = authServers.map { it.server.lowercase() }.toSet()
        val conflict = backendServers.firstOrNull { it.server.lowercase() in authNames }
        if (conflict != null) {
            throw IllegalArgumentException(
                "Server '${conflict.server}' is configured as both authentication and backend server. " +
                        "Move it to only one server group."
            )
        }
        if (forcedHosts.entries.any { it.key.isBlank() || it.value.isBlank() }) {
            throw IllegalArgumentException("servers.forced-hosts must not contain blank hostnames or server names")
        }
    }

    /** First auth target retained as a convenience property for older platform adapters. */
    val authServer: String
        get() = authServers.firstOrNull()?.server.orEmpty()

    /** First backend target retained as a convenience property for older platform adapters. */
    val backendServer: String
        get() = backendServers.firstOrNull()?.server.orEmpty()

    val maxPlayersPerServer: Int
        get() = 100

    val serverLimits: Map<String, Int>
        get() = (authServers + backendServers).associate { it.server to it.maxOnline }

    fun getEffectiveBackendServers(): List<String> = backendServers.map { it.server }

    fun getEffectiveAuthServers(): List<String> = authServers.map { it.server }

    fun hasBackendServer(): Boolean = backendServers.isNotEmpty()

    fun isAuthServer(serverName: String): Boolean =
        authServers.any { it.server.equals(serverName, ignoreCase = true) }

    fun requiringServerAuth(): ProxySettings =
        if (requireServerAuth) this else copy(requireServerAuth = true)

    companion object {
        @JvmStatic
        fun defaults(): ProxySettings = ProxySettings(
            true,
            listOf(ServerTarget("auth", 100)),
            listOf(ServerTarget("hub", 100)),
            emptyMap()
        )

        private fun effectiveNames(primary: String, values: List<String>): List<String> {
            val clean = values.map { it.trim() }.filter { it.isNotEmpty() }.distinct()
            if (clean.isNotEmpty()) return clean
            return primary.trim().takeIf { it.isNotEmpty() }?.let(::listOf) ?: emptyList()
        }
    }
}
